crowbar_framework/lib/upgrade.rb

```diff
@@ -0,0 +1,64 @@
+class Upgrade
+  attr_accessor :current_step, :substep, :current_node, :upgrade_progress
+  attr_reader :steps
+
+  # Return the current state of upgrade process.
+  # We're keeping the information in the file so is accessible by other applications
+  # and different crowbar versions.
+  def initialize
+    @upgrade_progress = {
+      current_step: upgrade_steps_6_7.first,
+      # substep is needed for more complex steps like upgrading the nodes
+      substep: "",
+      # current node is relevant only for the nodes_upgrade step
+      current_node: ""
+    }
+    if progress_file_path.exist?
+      @upgrade_progress = JSON.load(progress_file_path.read)
```

Comment:
Writing from two different objects would overwrite the status, yes. But how do you want to prevent it, reading the file each time we're accessing the attribute? That means more file accesses. Or going back to class method only?